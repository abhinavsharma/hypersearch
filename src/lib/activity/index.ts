/**
 * @module lib:activity
 * @version 1.0.0
 * @license (C) Insight
 */

import {
  MAX_INACTIVE_SECONDS,
  TRIGGER_START_TRACK_TIMER_MESSAGE,
  TRIGGER_STOP_TRACK_TIMER_MESSAGE,
} from 'constant';
import { getPublicationUrl } from 'lib/helpers';

export const activityMonitor = (document: Document): void => {
  if (!getPublicationUrl(window.location.href)) {
    return;
  }
  let cancelled = false;
  let secondsSinceLastActivity = 0;

  const handleMonitor = () => {
    secondsSinceLastActivity++;
    if (secondsSinceLastActivity > MAX_INACTIVE_SECONDS) {
      chrome.runtime.sendMessage({
        type: TRIGGER_STOP_TRACK_TIMER_MESSAGE,
        url: window.location.href,
      });
      cancelled = true;
      clearInterval(monitor);
    }
  };

  let monitor = setInterval(handleMonitor, 1000);

  const activity = () => {
    if (cancelled) {
      cancelled = false;
      chrome.runtime.sendMessage({
        type: TRIGGER_START_TRACK_TIMER_MESSAGE,
        url: window.location.href,
      });
      monitor = setInterval(handleMonitor, 1000);
    }
    secondsSinceLastActivity = 0;
  };

  const activityEvents = ['mousedown', 'mousemove', 'keydown', 'scroll'];
  activityEvents.forEach(function (eventName) {
    document.addEventListener(eventName, activity, true);
  });
};
