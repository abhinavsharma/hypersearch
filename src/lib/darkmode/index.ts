/**
 * @module lib:darkmode
 * @license (C) Insight
 * @version 1.0.0
 */

const Darkmode = require('./darkmode');

class DarkMode {

  public enable = async (document: Document) => {
    const darkmode = Darkmode(document);
    darkmode.auto(undefined, undefined, true);
  };

}

/**
 * Static instance of the darkmode.
 */
 const instance = new DarkMode();

 export default instance;
