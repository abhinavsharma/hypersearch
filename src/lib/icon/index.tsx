import React from 'react';
import { library } from '@fortawesome/fontawesome-svg-core';
import { fas } from '@fortawesome/free-solid-svg-icons';
import { far } from '@fortawesome/free-regular-svg-icons';
import { fab } from '@fortawesome/free-brands-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import '@fortawesome/fontawesome/styles.css';

library.add(fas);
library.add(far);
library.add(fab);

const faStyle = (style?: string) => {
  if (style === 'regular') {
    return 'far';
  } else if (style === 'brands') {
    return 'fab';
  }

  return 'fas';
};

const fontForIcon = (icon: AugmentationIcon) => {
  if (icon.font === 'font-awesome') {
    const type = faStyle(icon.style);
    return <FontAwesomeIcon icon={[ type, icon.name as any ]} />;
  }
}

export const handleIcon = (icon?: AugmentationIcon, emoji?: string, fallbackElement?: any) => {
  if (icon) {
    return fontForIcon(icon);
  }

  if (fallbackElement) {
    const FallbackElement = fallbackElement;
    return <FallbackElement />;
  } else {
    return emoji;
  }
};
