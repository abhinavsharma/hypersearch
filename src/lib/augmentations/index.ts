/**
 * @module lib:augmentations
 * @version 1.0.0
 * @license (C) Insight
 */

import { v4 as uuid } from 'uuid';
import SidebarLoader from 'lib/sidebar';
import SearchEngineManager from 'lib/engines';
import UserManager from 'lib/user';
import { debug, removeProtocol, encodeSpace, extractUrlProperties } from 'lib/helpers';
import {
  EXTENSION_SHARE_URL,
  NUM_DOMAINS_TO_EXCLUDE,
  UPDATE_SIDEBAR_TABS_MESSAGE,
  IGNORED_PREFIX,
  INSTALLED_PREFIX,
  PINNED_PREFIX,
  EXTENSION_BLOCKLIST_ADD_DOMAIN,
  EXTENSION_BLOCKLIST_REMOVE_DOMAIN,
  EXTENSION_AUGMENTATION_SAVE,
  MY_TRUSTLIST_TEMPLATE,
  DEDICATED_SERP_REGEX,
  ACTION_KEY,
  ACTION_LABEL,
  LEGACY_CONDITION_TYPE,
  CONDITION_KEY,
  AUGMENTATION_ID,
  DISABLED_KEYS,
  LUMOS_API_URL,
  ENV,
} from 'constant';

class AugmentationManager {
  /**
   * Helper property to track when to log the save action of an installed augmentation.
   */
  public preparedLogMessage: Record<'augmentation', Augmentation> | null;

  constructor() {
    debug('AugmentationManager - initialize\n---\n\tSingleton Instance', this, '\n---');
    this.preparedLogMessage = Object.create(null);
  }

  /**
   * Decide whether an augmentation is valid and can be used in the application.
   *
   * @param augmentation - The augmentation object
   * @returns boolean
   */
  public isAugmentationEnabled(augmentation: Augmentation) {
    const operations: Array<ActionObject | ConditionObject> = [
      ...augmentation.conditions.condition_list,
      ...augmentation.actions.action_list,
    ];

    return operations?.every((operation: ConditionObject | ActionObject) => {
      const key = (operation as ConditionObject).unique_key ?? operation.key;
      const hasValidKey = !(DISABLED_KEYS as readonly string[]).includes(key);
      return hasValidKey;
    });
  }

  /**
   * Add a domain (or publication) to the current user's blocklist
   *
   * @param domain - The domain to add
   */
  public async updateBlockList(domain: string) {
    const blockList = SidebarLoader.installedAugmentations
      .concat(SidebarLoader.otherAugmentations)
      .find(({ id }) => id === AUGMENTATION_ID.BLOCKLIST) as Augmentation;
    const isNewBlock = !blockList.actions.action_list.find(({ value }) => value[0] === domain);
    const newActionList: ActionObject[] = [
      ...blockList.actions.action_list.filter(
        (action) => action.value.length && !action.value.includes(domain),
      ),
      {
        key: ACTION_KEY.SEARCH_HIDE_DOMAIN,
        label: ACTION_LABEL.SEARCH_HIDE_DOMAIN,
        type: LEGACY_CONDITION_TYPE.LIST,
        value: [domain],
      },
    ];
    blockList.actions.action_list = newActionList;
    isNewBlock &&
      !UserManager.user.privacy &&
      SidebarLoader.sendLogMessage(EXTENSION_BLOCKLIST_ADD_DOMAIN, {
        domain,
      });
    this.addOrEditAugmentation(blockList, {
      actions: newActionList,
    });

    this.recordTrustBlock('BLOCK', domain, true);
  }

  /**
   * Remove a domain (or publication) from the current user's blocklist
   *
   * @param domain - The domain to remove
   */
  public async deleteFromBlockList(domain: string) {
    const blockList = SidebarLoader.installedAugmentations
      .concat(SidebarLoader.otherAugmentations)
      .find(({ id }) => id === AUGMENTATION_ID.BLOCKLIST) as Augmentation;
    const newActionList = [
      ...blockList.actions.action_list.filter(({ key, value }) =>
        key === ACTION_KEY.SEARCH_HIDE_DOMAIN ? value[0] !== domain : true,
      ),
    ];
    blockList.actions.action_list = newActionList;
    this.addOrEditAugmentation(blockList, {
      actions: newActionList,
    }),
      (SidebarLoader.hideDomains = SidebarLoader.hideDomains.filter((hidden) => hidden !== domain));
    !UserManager.user.privacy &&
      SidebarLoader.sendLogMessage(EXTENSION_BLOCKLIST_REMOVE_DOMAIN, {
        domain,
      });

    this.recordTrustBlock('BLOCK', domain, false);
  }

  /**
   * Toggle marking a domain as a trusted source.
   *
   * @param domain - The domain to be added or removed from the trusted sources
   */
  public async toggleTrustlist(domain: string) {
    const trustList = SidebarLoader.installedAugmentations
      .concat(SidebarLoader.otherAugmentations)
      .find(({ id }) => id === AUGMENTATION_ID.TRUSTLIST) as Augmentation;
    const existingDomain = !!(trustList.actions.action_list[0]?.value ?? []).includes(domain);
    const newActionValue = existingDomain
      ? (trustList.actions.action_list[0]?.value ?? []).filter((value) => value !== domain)
      : (trustList.actions.action_list[0]?.value ?? []).concat(domain);
    this.addOrEditAugmentation(trustList, {
      actions: [
        {
          ...(trustList.actions.action_list[0] ?? MY_TRUSTLIST_TEMPLATE.actions.action_list[0]),
          value: newActionValue,
        },
      ],
    });

    this.recordTrustBlock('TRUST', domain, !existingDomain);
  }

  /**
   * Transpose special characters in {@link ACTION_KEY.OPEN_URL open page} actions.
   * > Specification: https://bit.ly/3rUCl5d
   *
   * @param actionString - The action's value'
   * @param regexGroups - The possible RegExp groups from conditions
   * @returns The generated URL
   */
  public processOpenPageActionString(actionString: string, regexGroups: string[] = []) {
    let url = `https://${removeProtocol(actionString)}`;
    if (actionString.search(/%s[^r]+|%s$/gi) > -1) {
      url = url.replace('%s', encodeSpace(SidebarLoader.query));
    }
    if (actionString.indexOf('%u') > -1) url = url.replace('%u', SidebarLoader.url.href);
    if (actionString.search(/%sr[\d]{1,}/gi) > -1) {
      const domainIndices = url.match(/%sr[\d]*/gi) ?? [];
      domainIndices.forEach((actionString) => {
        const index = actionString.split('%sr')[1];
        url = url.replaceAll(
          `%sr${index}`,
          SidebarLoader.publicationSlices['original']?.[Number(index) - 1],
        );
      });
    }
    if (actionString.search(/%m[\d]{1,}/gi) > -1) {
      const domainIndices = url.match(/%m[\d]*/gi) ?? [];
      domainIndices.forEach((actionString) => {
        const index = actionString.split('%m')[1];
        url = url.replaceAll(`%m${index}`, regexGroups[Number(index) - 1]);
      });
    }
    return new URL(url);
  }

  /**
   * Create a new tab searching the same {@link SidebarLoader.query query } but using the specified custom engine.
   *
   * @param engine - The search engine object
   * @returns
   */
  public processSearchAlsoActionString(engine: SearchEngineObject['search_engine_json']) {
    const url = new URL(`https://${removeProtocol(engine.required_prefix)}`);
    engine.required_params.forEach((param) => url.searchParams.append(param, SidebarLoader.query));
    return url;
  }

  /**
   * Mark a suggested augmentation as hidden from the {@link SidebarLoader.fetchSubtabs subtabs response}.
   *
   * @param augmentation - The suggested augmentation to disable
   */
  public disableSuggestedAugmentation(augmentation: Augmentation) {
    SidebarLoader.ignoredAugmentations.push(augmentation);
    chrome.storage.local.set({
      [`${IGNORED_PREFIX}-${augmentation.id}`]: augmentation,
    });
    SidebarLoader.pinnedAugmentations = SidebarLoader.pinnedAugmentations.filter(
      (i) => i.id !== augmentation.id,
    );
    SidebarLoader.suggestedAugmentations = SidebarLoader.suggestedAugmentations.filter(
      (i) => i.id !== augmentation.id,
    );
    chrome.runtime.sendMessage({ type: UPDATE_SIDEBAR_TABS_MESSAGE });
  }

  /**
   * Remove hide marker from a suggested augmentation from the {@link SidebarLoader.fetchSubtabs subtabs response}.
   *
   * @param augmentation - The suggested augmentation to enable
   */
  public enableSuggestedAugmentation(augmentation: Augmentation) {
    SidebarLoader.ignoredAugmentations = SidebarLoader.ignoredAugmentations.filter(
      (i) => i.id !== augmentation.id,
    );
    chrome.storage.local.remove(`${IGNORED_PREFIX}-${augmentation.id}`);
    const { isRelevant } = this.getAugmentationRelevancy(augmentation);
    // prettier-ignore
    augmentation.pinned
      ? SidebarLoader.pinnedAugmentations.push(augmentation)
      : isRelevant
        ? SidebarLoader.suggestedAugmentations.push(augmentation)
        : SidebarLoader.otherAugmentations.push(augmentation);
    chrome.runtime.sendMessage({ type: UPDATE_SIDEBAR_TABS_MESSAGE });
  }

  /**
   * Mark an augmentation to be always visible in the Sidebar.
   *
   * @param augmentation - The augmentation to pin
   */
  public pinAugmentation(augmentation: Augmentation) {
    augmentation.pinned = true;
    SidebarLoader.pinnedAugmentations.push(augmentation);
    chrome.storage.local.set({
      [`${PINNED_PREFIX}-${augmentation.id}`]: augmentation,
    });
    if (
      !SidebarLoader.installedAugmentations.find(({ id }) => id === augmentation.id) &&
      !SidebarLoader.suggestedAugmentations.find(({ id }) => id === augmentation.id)
    ) {
      SidebarLoader.enabledOtherAugmentations.push(augmentation);
    }
    chrome.runtime.sendMessage({ type: UPDATE_SIDEBAR_TABS_MESSAGE });
  }

  /**
   * Remove an augmentation from the pinned list.
   *
   * @param augmentation - The augmentation to unpin
   */
  public unpinAugmentation(augmentation: Augmentation) {
    augmentation.pinned = false;
    SidebarLoader.pinnedAugmentations = SidebarLoader.pinnedAugmentations.filter(
      (i) => i.id !== augmentation.id,
    );
    SidebarLoader.enabledOtherAugmentations = SidebarLoader.enabledOtherAugmentations.filter(
      ({ id }) => id !== augmentation.id,
    );
    chrome.storage.local.remove(`${PINNED_PREFIX}-${augmentation.id}`);
    chrome.runtime.sendMessage({ type: UPDATE_SIDEBAR_TABS_MESSAGE });
  }

  /**
   * Permanently delete a custom user created augmentation from both the Sidebar and storage.
   *
   * @param augmentation - The installed augmentation to remove
   */
  public removeInstalledAugmentation(augmentation: Augmentation) {
    this.unpinAugmentation(augmentation);
    SidebarLoader.pinnedAugmentations = SidebarLoader.pinnedAugmentations.filter(
      (i) => i.id !== augmentation.id,
    );
    SidebarLoader.installedAugmentations = SidebarLoader.installedAugmentations.filter(
      (i) => i.id !== augmentation.id,
    );
    SidebarLoader.otherAugmentations = SidebarLoader.otherAugmentations.filter(
      (i) => i.id !== augmentation.id,
    );
    chrome.storage.local.remove(augmentation.id);
    chrome.runtime.sendMessage({ type: UPDATE_SIDEBAR_TABS_MESSAGE });
  }

  /**
   * Decide if the augmentation's conditions evaluate to match the current page.
   *
   * @param augmentation - The augmentation to check
   * @returns - {@link AugmentationRelevancyResult}
   */
  public getAugmentationRelevancy(augmentation: Augmentation): AugmentationRelevancyResult {
    if (!augmentation?.actions || !augmentation.conditions) {
      return Object.create(null);
    }

    // ! SEARCH DOMAINS
    const hasAnyPageCondition = !!augmentation.conditions.condition_list.filter(
      ({ key, unique_key }) =>
        ((unique_key ?? key) === CONDITION_KEY.ANY_SEARCH_ENGINE &&
          SidebarLoader.isSerp &&
          SidebarLoader.url.href.match(DEDICATED_SERP_REGEX)) ||
        (unique_key ?? key) === CONDITION_KEY.ANY_URL,
    ).length;

    const domainsToLookCondition =
      augmentation.conditions.condition_list.reduce(
        (conditions, { key, unique_key, value }) =>
          (unique_key ?? key) === CONDITION_KEY.SEARCH_CONTAINS
            ? conditions.concat(value as unknown as string)
            : conditions,
        [] as string[],
      ) ?? [];

    const domainsToLookAction =
      augmentation.actions?.action_list.reduce(
        (actions, { key, value }) =>
          key === ACTION_KEY.SEARCH_DOMAINS ? actions.concat(value as unknown as string) : actions,
        [] as string[],
      ) ?? [];

    const matchingDomainsCondition =
      SidebarLoader.domains?.filter((value) =>
        domainsToLookCondition?.find((i) => value?.search(new RegExp(`^${i}`, 'gi')) > -1),
      ) ?? [];

    const matchingDomainsAction =
      SidebarLoader.domains?.filter((value) =>
        domainsToLookAction?.find((i) => value?.search(new RegExp(`^${i}`, 'gi')) > -1),
      ) ?? [];

    const matchingDomains =
      // must have at least one matching domain
      matchingDomainsCondition
        .map(
          (domain) =>
            !!SidebarLoader.domains?.find((e) => e?.search(new RegExp(`^${domain}`, 'gi')) > -1),
        )
        .filter((isMatch) => !!isMatch).length > 0 &&
      // exclude if too much overlap
      matchingDomainsAction
        .map(
          (domain) =>
            !!SidebarLoader.domains?.find((e) => e?.search(new RegExp(`^${domain}`, 'gi')) > -1),
        )
        .filter((isMatch) => !!isMatch).length <
        (hasAnyPageCondition || !!SidebarLoader.url.searchParams.get('insight-tour')
          ? Infinity
          : NUM_DOMAINS_TO_EXCLUDE);

    // ! SEARCH QUERY
    const matchingQuery = augmentation.conditions?.condition_list.some(
      ({ key, unique_key, value }) =>
        (unique_key ?? key) === CONDITION_KEY.SEARCH_QUERY_CONTAINS &&
        SidebarLoader.query?.search(value[0] as string) > -1,
    );

    // ! SEARCH INTENT
    let hasPreventAutoexpand = false;
    const matchingIntent = augmentation.conditions.condition_list
      .reduce((intents, { key, unique_key, value }) => {
        if ((unique_key ?? key) === CONDITION_KEY.SEARCH_INTENT_IS) {
          const matchingIntent = SearchEngineManager.intents.find(
            ({ intent_id }) => intent_id === value[0],
          );
          if (matchingIntent?.stay_collapsed) {
            hasPreventAutoexpand = true;
          }
          if (matchingIntent) {
            // matching intent domains
            const intentDomains = matchingIntent.sites.split(',') ?? [];
            intentDomains.forEach(
              (domain) =>
                !!SidebarLoader.publicationSlices['original']?.find(
                  (mainSerpDomain) => !!mainSerpDomain.match(domain)?.length,
                ) && intents.push(domain),
            );
            // matching intent elements on document
            if (matchingIntent.google_css) {
              intents.concat(Array.from(document.querySelectorAll(matchingIntent.google_css)));
            }
          }
        }
        return intents;
      }, [] as Array<string | Element>)
      .filter((isMatch) => !!isMatch);

    // ! SEARCH ENGINE
    const matchingEngine = !!augmentation.conditions.condition_list.find(
      ({ key, unique_key, value }) => {
        if ((unique_key ?? key) === CONDITION_KEY.SEARCH_ENGINE_IS) {
          const cse = value[0] as unknown as SearchEngineObject;
          const hasAllMatchingParams = (cse.search_engine_json ?? cse)?.required_params?.every(
            (param) => !!SidebarLoader.url.searchParams.get(param),
          );
          const hasRequiredPrefix =
            SidebarLoader.url.href.search((cse.search_engine_json ?? cse)?.required_prefix) > -1;
          return hasAllMatchingParams && hasRequiredPrefix;
        }
        return null;
      },
    );

    // ! URL/DOMAIN MATCH
    let numRegexConditions = 0;
    const regexConditions: KeyEventMap<ConditionKey> = [
      CONDITION_KEY.URL_EQUALS,
      CONDITION_KEY.URL_MATCHES,
      CONDITION_KEY.DOMAIN_EQUALS,
      CONDITION_KEY.DOMAIN_MATCHES,
      CONDITION_KEY.DOMAIN_CONTAINS,
    ];
    const matchingUrl = augmentation.conditions.condition_list.reduce((matches, condition) => {
      const { unique_key: key, value } = condition;
      if (!key) {
        return [];
      }
      if (regexConditions.includes(key)) {
        numRegexConditions += 1;
      }
      if (key === CONDITION_KEY.URL_EQUALS && SidebarLoader.url.href === value[0]) {
        matches.push(true);
      }
      if (
        key === CONDITION_KEY.URL_MATCHES &&
        SidebarLoader.url.href.match(value[0] as string)?.length
      ) {
        matches.push(true);
      }
      if (
        key === CONDITION_KEY.DOMAIN_EQUALS &&
        extractUrlProperties(SidebarLoader.url.href).hostname === value[0]
      ) {
        matches.push(true);
      }
      if (
        key === CONDITION_KEY.DOMAIN_MATCHES &&
        extractUrlProperties(SidebarLoader.url.href).hostname?.match(value[0] as string)?.length
      ) {
        matches.push(true);
      }
      if (
        key === CONDITION_KEY.DOMAIN_CONTAINS &&
        value.includes(extractUrlProperties(SidebarLoader.url.href).hostname ?? '')
      ) {
        matches.push(true);
      }
      return matches;
    }, [] as boolean[]);

    const isMatchingNote = !!augmentation.actions.action_list.find(({ key, value }) => {
      key === ACTION_KEY.URL_NOTE && SidebarLoader.url.href.includes(value[0] as string);
    });

    const evaluationMatch =
      augmentation.conditions.evaluate_with === 'AND'
        ? augmentation.conditions.condition_list?.every(({ key, unique_key }) => {
            const actualKey = unique_key ?? key;
            switch (actualKey) {
              case CONDITION_KEY.SEARCH_CONTAINS:
                return matchingDomains;
              case CONDITION_KEY.SEARCH_QUERY_CONTAINS:
                return matchingQuery;
              case CONDITION_KEY.SEARCH_ENGINE_IS:
                return matchingEngine;
              case CONDITION_KEY.SEARCH_INTENT_IS:
                return matchingIntent;
              case CONDITION_KEY.ANY_SEARCH_ENGINE:
              case CONDITION_KEY.ANY_URL:
                return hasAnyPageCondition;
              case CONDITION_KEY.URL_EQUALS:
              case CONDITION_KEY.URL_MATCHES:
              case CONDITION_KEY.DOMAIN_EQUALS:
              case CONDITION_KEY.DOMAIN_MATCHES:
                return numRegexConditions === matchingUrl.length;
            }
          })
        : true;

    return {
      isHidden: !!augmentation.installed && !augmentation.enabled,
      isRelevant:
        (augmentation.enabled || !augmentation.installed) &&
        (augmentation.pinned ||
          (evaluationMatch &&
            (hasAnyPageCondition ||
              matchingQuery ||
              matchingDomains ||
              !!matchingIntent.length ||
              matchingEngine ||
              !!matchingUrl.length ||
              isMatchingNote))),
      matchingIntent,
      hasPreventAutoexpand,
      domainsToLookAction,
      domainsToLookCondition,
      matchingDomainsAction,
      matchingDomainsCondition,
    };
  }

  /**
   * Send request to store an encoded (stringified) augmentation on the {@link EXTENSION_SHARE_URL extension site}
   *
   * @param encoded - The encoded augmentation
   */
  public async shareAugmentation(encoded: string) {
    await fetch(`${EXTENSION_SHARE_URL}${encodeURIComponent(encoded)}`, {
      mode: 'no-cors',
    });
  }

  /**
   * Create or edit an existing augmentation by the specified options.
   *
   * @param augmentation - The augmentation to add or edit
   * @param options - {@link AugmentationData}
   */
  public addOrEditAugmentation(
    augmentation: Augmentation,
    {
      actions,
      conditions,
      conditionEvaluation,
      description,
      name,
      isActive,
      isPinning,
    }: AugmentationData,
  ) {
    const customId = `${INSTALLED_PREFIX}-${
      augmentation.id !== '' ? augmentation.id : name?.replace(/[\s]/g, '_').toLowerCase()
    }-${uuid()}`;
    const id =
      augmentation.id.startsWith(`${INSTALLED_PREFIX}-`) && !isPinning ? augmentation.id : customId;
    const updated = {
      ...augmentation,
      id,
      name: name ?? augmentation.name,
      description: description ?? augmentation.description,
      conditions: {
        condition_list: conditions ?? augmentation.conditions.condition_list,
        evaluate_with: conditionEvaluation ?? augmentation.conditions.evaluate_with,
      },
      actions: {
        ...augmentation.actions,
        action_list: actions
          ? actions
              .map((action) => ({
                ...action,
                value: action.value.filter((isValue) => !!isValue),
              }))
              .filter(({ value }) => !!value.length)
          : augmentation.actions.action_list.filter(({ value }) => !!value.length),
      },
      enabled: isActive ?? augmentation.enabled,
      installed: true,
    };
    chrome.storage.local.set({ [id]: updated });

    const { isRelevant, isHidden } = this.getAugmentationRelevancy(updated);

    if (isRelevant) {
      SidebarLoader.installedAugmentations = [
        updated,
        ...SidebarLoader.installedAugmentations.filter((i) => i.id !== updated.id),
      ];
      SidebarLoader.otherAugmentations = SidebarLoader.otherAugmentations.filter(
        (i) => i.id !== updated.id,
      );
      SidebarLoader.ignoredAugmentations = SidebarLoader.ignoredAugmentations.filter(
        (i) => i.id !== updated.id,
      );
    } else {
      SidebarLoader.installedAugmentations = SidebarLoader.installedAugmentations.filter(
        (i) => i.id !== updated.id,
      );
      if (isHidden) {
        SidebarLoader.ignoredAugmentations = [
          updated,
          ...SidebarLoader.ignoredAugmentations.filter((i) => i.id !== updated.id),
        ];
      } else {
        SidebarLoader.otherAugmentations = [
          updated,
          ...SidebarLoader.otherAugmentations.filter((i) => i.id !== updated.id),
        ];
      }
    }
    this.preparedLogMessage &&
      !UserManager.user.privacy &&
      SidebarLoader.sendLogMessage(EXTENSION_AUGMENTATION_SAVE, {
        augmentation: this.preparedLogMessage.augmentation,
      });
    chrome.runtime.sendMessage({ type: UPDATE_SIDEBAR_TABS_MESSAGE });
    debug(
      'AugmentationManager - addOrEditAugmentation\n---\n\tOriginal',
      augmentation,
      '\n\tUpdated',
      updated,
      '\n---',
    );
  }

  private recordTrustBlock = async (type: 'TRUST' | 'BLOCK', domain: string, toggle: boolean) => {
    const authorization = await UserManager.getAccessToken();

    if (UserManager.user.privacy !== false || !authorization) {
      return;
    }

    const endpoint = type === 'TRUST' ? 'trust' : 'block';
    const body = {
      url: domain,
      toggle,
    };
    await fetch(`${LUMOS_API_URL[ENV]}toggleList/${endpoint}`, {
      method: 'POST',
      headers: {
        authorization,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  };
}

/**
 * Static instance of the augmentation manager.
 */
const instance = new AugmentationManager();

export default instance;
