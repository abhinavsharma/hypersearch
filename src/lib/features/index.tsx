/**
 * @module lib:features
 * @version 1.0.0
 * @license (C) Insight
 */

import React, { useCallback, useEffect, useState } from 'react';
import SidebarLoader from 'lib/sidebar';
import { triggerSerpProcessing } from 'lib/helpers';
import { DEV_FEATURE_FLAGS, FEATURE_FLAG_BLOB_URL, UPDATE_SIDEBAR_TABS_MESSAGE } from 'constant';

/**
 * Hook for getting the current state of the specified feature.
 *
 * @param feature - The name of the feature
 * @returns [Status, Setter]
 */
export const useFeature = (feature: string) => {
  const [localFlags, setLocalFlags] = useState<Features>(Object.create(null));
  const [features, setFeatures] = useState<Features>(Object.create(null));

  const getFeatures = useCallback(async () => {
    const remoteBlob = await fetch(FEATURE_FLAG_BLOB_URL, { mode: 'cors' });
    const raw = await remoteBlob.json();
    const locals = await new Promise<Record<string, Features>>((resolve) =>
      chrome.storage.local.get(DEV_FEATURE_FLAGS, resolve),
    ).then((data) => data[DEV_FEATURE_FLAGS]);
    setLocalFlags(locals);
    setFeatures(
      Object.assign(
        raw['features'].reduce((record: Features, entry: FeatureEntry) => {
          record[entry.name] = entry.enabled;
          return record;
        }, Object.create(null)),
        locals,
      ),
    );
  }, []);

  useEffect(() => {
    getFeatures();
  }, [getFeatures]);

  return [
    features[feature],
    () => {
      chrome.storage.local.set({
        [DEV_FEATURE_FLAGS]: {
          ...localFlags,
          [feature]: !features[feature],
        },
      });
      setFeatures({ ...features, [feature]: !features[feature] });
      chrome.runtime.sendMessage({ type: UPDATE_SIDEBAR_TABS_MESSAGE });
      triggerSerpProcessing(SidebarLoader, false);
    },
  ] as [boolean, () => void];
};

export const FeatureGate: FeatureGate = ({ component, feature, fallback }) => {
  const [enabled] = useFeature(feature);
  return <>{enabled ? component : fallback ?? null}</>;
};
