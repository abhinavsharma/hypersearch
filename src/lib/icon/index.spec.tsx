import React from 'react';
import { handleIcon } from 'lib/icon';
import renderer from 'react-test-renderer';

describe('Icon tests', () => {

  describe('handleIcon', () => {

    test('With font-awesome fas', () => {
      // Given
      const icon = handleIcon({
        name: 'circle',
        font: 'font-awesome',
        style: 'solid',
      });

      // When
      const component = renderer.create(
        <span>{ icon }</span>
      );

      const tree = component.toJSON()!;

      // Then
      expect(tree).toMatchSnapshot();
    });

    test('With font-awesome fab', () => {
      // Given
      const icon = handleIcon({
        name: 'apple',
        font: 'font-awesome',
        style: 'brands',
      });

      // When
      const component = renderer.create(
        <span>{ icon }</span>
      );

      const tree = component.toJSON()!;

      // Then
      expect(tree).toMatchSnapshot();
    });

    test('With font-awesome far', () => {
      // Given
      const icon = handleIcon({
        name: 'circle',
        font: 'font-awesome',
        style: 'regular',
      });

      // When
      const component = renderer.create(
        <span>{ icon }</span>
      );

      const tree = component.toJSON()!;

      // Then
      expect(tree).toMatchSnapshot();
    });

    test('Should fallback to emoji if icon not informed', () => {
      // Given
      // Initial state

      // When
      const fallback = handleIcon(undefined, 'ðŸ˜„');

      // Then
      expect(typeof fallback).toBe('string');
      expect(fallback).toBe('ðŸ˜„');
    });

    test('Should fallback to emoji if icon not informed', () => {
      // Given
      const Element = () => {
        return <h1></h1>;
      }

      // When
      const fallback = handleIcon(undefined, undefined, Element);

      // Then
      expect(typeof fallback).toBe('object');
      expect(fallback).toStrictEqual(<Element />);
    });

  });

})
