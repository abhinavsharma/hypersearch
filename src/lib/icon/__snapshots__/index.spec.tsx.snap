// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon tests handleIcon With font-awesome fab 1`] = `
<span>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-apple "
    data-icon="apple"
    data-prefix="fab"
    focusable="false"
    role="img"
    style={Object {}}
    viewBox="0 0 384 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M318.7 268.7c-.2-36.7 16.4-64.4 50-84.8-18.8-26.9-47.2-41.7-84.7-44.6-35.5-2.8-74.3 20.7-88.5 20.7-15 0-49.4-19.7-76.4-19.7C63.3 141.2 4 184.8 4 273.5q0 39.3 14.4 81.2c12.8 36.7 59 126.7 107.2 125.2 25.2-.6 43-17.9 75.8-17.9 31.8 0 48.3 17.9 76.4 17.9 48.6-.7 90.4-82.5 102.6-119.3-65.2-30.7-61.7-90-61.7-91.9zm-56.6-164.2c27.3-32.4 24.8-61.9 24-72.5-24.1 1.4-52 16.4-67.9 34.9-17.5 19.8-27.8 44.3-25.6 71.9 26.1 2 49.9-11.4 69.5-34.3z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</span>
`;

exports[`Icon tests handleIcon With font-awesome far 1`] = `
<span>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-circle "
    data-icon="circle"
    data-prefix="far"
    focusable="false"
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 256C512 397.4 397.4 512 256 512C114.6 512 0 397.4 0 256C0 114.6 114.6 0 256 0C397.4 0 512 114.6 512 256zM256 48C141.1 48 48 141.1 48 256C48 370.9 141.1 464 256 464C370.9 464 464 370.9 464 256C464 141.1 370.9 48 256 48z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</span>
`;

exports[`Icon tests handleIcon With font-awesome fas 1`] = `
<span>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-circle "
    data-icon="circle"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 256C512 397.4 397.4 512 256 512C114.6 512 0 397.4 0 256C0 114.6 114.6 0 256 0C397.4 0 512 114.6 512 256z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</span>
`;
